Support for progressive AVIFs and operating point selection

joedrago opened the pull request:
* Support for parsing boxes: a1op, a1lx, lsel
* Added operating point and "skip" support to avifCodec impls for layer selection support
* Reworked avifdec decoding to support nth image decoding and minor optimization
* New avifProgressiveState enum to expose info about progressive AVIFs to users of libavif
* New avifdec args: --progressive, --index

@wantehchang Do not merge this yet -- There are still outstanding issues with a few of the Apple example files. It is unclear if they are issues with the file itself (`animals_00_multilayer_grid_a1lx.avif`), or an issue with our usage of dav1d (`animals_00_multilayer.avif`).

joedrago commented:
Some open questions, pinging folks directly (@leo-barnes, @cconcolato):

For some reason, the example file `animals_00_multilayer.avif` refuses to decode the primary item's AV1 payload when I use dav1d, but decodes perfectly fine when I use libaom or libgav1. I'd consider it to be a bug in libavif if it wasn't the only file (out of all AVIFs I've ever seen) with this specific issue. It doesn't give any specific error, I just don't think it is "enough" data to output a frame, at least for dav1d. I could use some help here on this one.

`animals_00_multilayer_grid_a1lx.avif` is an entirely different question. This is offering a _progressive grid_ image, but the earlier layers are not large enough to "fill the grid", as they are half resolution to the "real" grid cells. I fear that the only way this could possibly work is if these smaller cells are expected to be scaled prior to being stitched into a grid, which seems like it is taking these transformations too far. We already put in limitations on AVIF's derived image transformations for grid. Perhaps this can be another (e.g. "progressive grid cell variants must use fixed dimensions").

joedrago commented:
I've diagnosed the issue somewhat with `animals_00_multilayer.avif`, but I don't know what to do with the discrepancy between `dav1d` and `libaom`'s behavior. dav1d has this block:

```
    // skip lower spatial layers
    if (c->operating_point_idc && !c->all_layers) {
        const int max_spatial_id = ulog2(c->operating_point_idc >> 8);
        if (max_spatial_id > c->out.frame_hdr->spatial_id) {
            dav1d_picture_unref_internal(&c->out);
            return 0;
        }
    }
```
![debug](https://user-images.githubusercontent.com/332342/119744517-0c184180-be41-11eb-88b2-c0a429f1b3fe.png)

This file's primary item does not signal any hint that it is multiple layers. In fact, it is one of the simplest types of primary items you can craft for AVIF (one extent, typical av1C/ispe/pixi/colr props, no alpha). Therefore, I don't configure any of the AV1 codec implementations (dav1d, aom, gav1) to have "all layers" turn on, as I just want this one AV1 payload's image.  As you can see above from the inspect variables and the logic here, dav1d is aware that there is another layer coming "later", so it discards the layer it has already encoded.

I feel like dav1d is doing the correct thing here and the file is invalid, but it happens to work just fine with libaom and libgav1, which I can't explain. I'm hoping either @leo-barnes or @wantehchang has some insight on this one.

leo-barnes commented:
@joedrago 
> As you can see above from the inspect variables and the logic here, dav1d is aware that there is another layer coming "later", so it discards the layer it has already decoded.

That can't be correct. It's fine if `dav1d` doesn't output the layer, but it can't discard it since future layers are predicted from it.

Either it's a bug in `dav1d`, or I somehow didn't correctly configure `libaom` when creating the two layers so it's not correctly signalling that the second layer is predicted from the first. @wantehchang, would it be possible for you to check if the second layer is correctly marked as using the first layer for prediction?

leo-barnes commented:
> animals_00_multilayer_grid_a1lx.avif is an entirely different question. This is offering a progressive grid image, but the earlier layers are not large enough to "fill the grid", as they are half resolution to the "real" grid cells. I fear that the only way this could possibly work is if these smaller cells are expected to be scaled prior to being stitched into a grid, which seems like it is taking these transformations too far. We already put in limitations on AVIF's derived image transformations for grid. Perhaps this can be another (e.g. "progressive grid cell variants must use fixed dimensions").

Spec-wise, this is perfectly valid:
```
NOTE: The dimensions of possible intermediate output images might not match the ones given in the ispe property. If they display these intermediate images, renderers are expected to scale the output image to match the ispe property.
```

If you want to support progressive refinement of grids, you will need to be prepared for intermediate layers having a different size than the tile size and will need to upscale them. If you don't want to do progressive refinement, you're guaranteed that the final layer (or the layer selected by `lsel`) will have the correct size.

joedrago commented:
I guess we should decide if we want AVIF implementations to be on the hook for YUV plane scaling.

wantehchang commented:
[Note: In this comment I use the variable names used in the AV1 spec to minimize the risk of confusion.]

I ran avifdec on `animals_00_multilayer.avif` in the debugger. The AV1 sequence header OBU has operating_points_cnt_minus_1=1 (which means there are two operating points) and operating_point_idc[0]=0x301 and operating_point_idc[1]=0x101.

The image item decoded by avifdec has only one AV1 frame OBU after the AV1 sequence header OBU. That AV1 frame OBU has obu_extension_flag=0, so it does not have an obu_extension_header() and its temporal_id and spatial_id are both 0 by default. Note: Under the draft revision of the AV1 spec, this AV1 frame OBU should set obu_extension_flag to 1 and have an obu_extension_header() with temporal_id=0 and spatial_id=0. It would be good to follow the new requirement.

The image item decoded by avifdec has two ispe properties, which violates the requirement in HEIF Section 6.5.3.1 that the quantity per item is one. The first ispe property has image_width=2048 and image_width=1536. The second ispe property has image_width=1024 and image_width=768.

Joe said `animals_00_multilayer.avif` has two image items. I think it should be changed to have only one image item. That image item should correspond to operatingPoint=0 and OperatingPointIdc=0x301 and have two AV1 Frame OBUs. And it should have only one ispe property.

Re: how an AV1 decoder should handle the current truncated AV1 bitstream in `animals_00_multilayer.avif`: I think the way dav1d handles this error condition is better.

joedrago commented:
@wantehchang It seems like the two outstanding issues I had (progressive grids and Leo's undefined-behavior file) are resolved sufficiently that you can review this at your leisure.

Remember that part of the most recent commit's design here (which uses `libyuv`'s `ScalePlane*()` functions) is that we would adopt/recontribute that small subset of functions into libavif, hopefully. Consider that as you review this, please.

wantehchang commented on apps/avifdec.c:
```diff
@@ -36,6 +36,9 @@ static void syntax(void)
     printf("                        (JPEG only; not applicable to y4m)\n");
+    printf("    --index           : When decoding an image sequence or progressive image, specify which frame index to decode (Default: 0)\n");
+    printf("    --progressive     : Enable progressive AVIF processing. If a progressive image is encountered, avifdec will use\n");
+    printf("                        --index to choose which variant to decode (in progressive order).\n");
```
Should we change "variant" to "layer" or "frame"?

joedrago replied:
Layer is probably better here.

wantehchang commented on apps/shared/avifutil.c:
```diff
@@ -105,17 +105,19 @@ static void avifImageDumpInternal(const avifImage * avif, uint32_t gridCols, uin
             printf("    * imir (Mirror)        : Mode %u (%s)\n", avif->imir.mode, (avif->imir.mode == 0) ? "top-to-bottom" : "left-to-right");
         }
     }
+    printf(" * Alpha          : %s\n", alphaPresent ? (avif->alphaPremultiplied ? "Premultiplied" : "Not premultiplied") : "Absent");
```
This line is the same as line 55 above. Is this line added by mistake?

joedrago replied:
Yes, fixing.

wantehchang commented on include/avif/avif.h:
```diff
@@ -772,6 +795,16 @@ typedef struct avifDecoder
     avifDecoderSource requestedSource;
 
+    // If this is true and a progressive AVIF is decoded, avifDecoder will behave as if the AVIF
+    // is an image sequence, in that it will set imageCount to the available number of progressive
```
Delete one of the two "available"s in "the available number of progressive frames available"

wantehchang commented on include/avif/avif.h:
```diff
@@ -772,6 +795,16 @@ typedef struct avifDecoder
+    // If this is true and a progressive AVIF is decoded, avifDecoder will behave as if the AVIF
+    // is an image sequence, in that it will set imageCount to the available number of progressive
+    // frames available, and avifDecoderNextImage()/avifDecoderNthImage() will all for specific
+    // variants of a progressive image to be decoded. To distinguish between a progressive AVIF
```
"will all for specific variants of a progressive image to be decoded" sounds strange. The word "all" seems incorrect, but I don't know what word it should be changed to.

joedrago replied:
It should be `will allow for specific layers`. Fixing.

wantehchang commented on include/avif/avif.h:
```diff
@@ -772,6 +795,16 @@ typedef struct avifDecoder
+    avifBool allowProgressive;
+
+    // --------------------------------------------------------------------------------------------
+    // Outputs
```
I like splitting the avifDecoder struct members into three sections: Inputs, Outputs, and Internals. But the current division is not quite right.

For example, in the Output sections, `ignoreExif,` `ignoreXMP`, `imageCountLimit`, and `strictFlags` are inputs, not outputs. Perhaps they should be moved to the Inputs section?

Also, in the Internals section, `diag` is not really internal. Should we move it to the Output section?

wantehchang commented on include/avif/internal.h:
```diff
@@ -244,6 +254,9 @@ typedef struct avifCodec
                                           //
     avifDiagnostics * diag;               // Shallow copy; owned by avifEncoder or avifDecoder
+                                          //
+    uint8_t operatingPointIndex;          // Operating point Index, defaults to 0.
```
I suggest renaming this member `operatingPoint`. , because this corresponds to the `operatingPoint` variable in the AV1 spec.

Note: AV1 spec also has an `operating_point_idc[]` array and an `OperatingPointIdc` variable. The current name of this member, `operatingPointIndex`,  may potentially be misinterpreted as the `OperatingPointIdc` variable in the AV1 spec.

wantehchang commented on src/avif.c:
```diff
@@ -101,6 +101,20 @@ const char * avifResultToString(avifResult result)
+            break;
+    }
+    // clang-format on
+    return "Unknown";
```
Move this return statement to the default case in the switch statement.

joedrago replied:
This is the same layout as `avifPixelFormatToString()`. If we want to change this pattern in all functions, let's do it in a separate PR.

wantehchang reviewed (changes requested):
Joe: I haven't reviewed the main changes in this PR. Here is the first round of review comments.

joedrago commented:
I've updated with most of the changes you requested. I left the default case return statement as it doesn't match other functions and we can adjust that style separately.

wantehchang commented on apps/avifdec.c:
```diff
@@ -36,6 +36,9 @@ static void syntax(void)
     printf("    -u,--upsampling U : Chroma upsampling (for 420/422). automatic (default), fastest, best, nearest, or bilinear\n");
     printf("    -r,--raw-color    : Output raw RGB values instead of multiplying by alpha when saving to opaque formats\n");
     printf("                        (JPEG only; not applicable to y4m)\n");
+    printf("    --index           : When decoding an image sequence or progressive image, specify which frame index to decode (Default: 0)\n");
```
For a progressive image, it seems that the default should be the best layer, which has the largest frame index. Is it easy to implement that?

joedrago replied:
If you don't pass `--progressive` at all, you will get the non-progressive behavior from the progressive AVIF, which will be the topmost/best layer.

wantehchang replied:
Thanks for the explanation. Perhaps in line 40 we can change

> If a progressive image is encountered,
to

> If a progressive image is encountered and --progressive is passed,

wantehchang commented on apps/avifdec.c:
```diff
@@ -178,17 +189,31 @@ int main(int argc, char * argv[])
+                       decoder->image->height);
+                ++currIndex;
+                nextImageResult = avifDecoderNextImage(decoder);
+            }
```
Lines 201-213 can be simplified slightly (only one avifDecoderNextImage call) if you are willing to do an assignment in the while loop's conditional expression:

```
            avifResult nextImageResult;
            while ((nextImageResult = avifDecoderNextImage(decoder)) == AVIF_RESULT_OK) {
                printf("   * Decoded frame [%d] [pts %2.2f (%" PRIu64 " timescales)] [duration %2.2f (%" PRIu64 " timescales)] [%ux%u]\n",
                       currIndex,
                       decoder->imageTiming.pts,
                       decoder->imageTiming.ptsInTimescales,
                       decoder->imageTiming.duration,
                       decoder->imageTiming.durationInTimescales,
                       decoder->image->width,
                       decoder->image->height);
                ++currIndex;
            }
```

wantehchang commented on apps/avifdec.c:
```diff
@@ -210,52 +235,73 @@ int main(int argc, char * argv[])
     } else {
-        printf("ERROR: Failed to decode image: %s\n", avifResultToString(decodeResult));
-        avifDumpDiagnostics(&decoder->diag);
+        fprintf(stderr, "Unrecognized file extension: %s\n", outputFilename);
```
Nit: If we reach here, we recognize the file extension (currently only .avif), but the file cannot be the input to avifenc. So this error message is a little confusing.

joedrago replied:
Perhaps "Unsupported input file extension" might be better wording here.

wantehchang commented on apps/avifdec.c:
```diff
@@ -210,52 +235,73 @@ int main(int argc, char * argv[])
-    avifImageDestroy(avif);
+
+cleanup:
+    if (decoder) {
```
This null pointer check for `decoder` is not necessary because `decoder` cannot be null. See line 238:

`avifDecoder * decoder = avifDecoderCreate();`

wantehchang commented on src/read.c:
```diff
@@ -37,6 +37,8 @@ static const size_t xmpContentTypeSize = sizeof(xmpContentType);
 // can't be more than 4 unique tuples right now.
 #define MAX_IPMA_VERSION_AND_FLAGS_SEEN 4
 
+#define MAX_AV1_LAYER_COUNT 4
```
Optional: Would be good to comment that 4 is the maximum number of spatial layers in AV1.

wantehchang commented on src/read.c:
```diff
@@ -361,11 +384,11 @@ static uint32_t avifGetSampleCountOfChunk(const avifSampleTableSampleToChunkArra
+static avifBool avifCodecDecodeInputFillFromSampleTable(avifCodecDecodeInput * decodeInput,
+                                                        avifSampleTable * sampleTable,
+                                                        const uint32_t imageCountLimit,
+                                                        const uint64_t sizeHint,
```
Nit/optional: Remove "const" from these two parameters.

wantehchang commented on src/read.c:
```diff
@@ -451,6 +475,119 @@ static avifBool avifCodecDecodeInputGetSamples(avifCodecDecodeInput * decodeInpu
+                                                        avifDecoderItem * item,
+                                                        avifBool allowProgressive,
+                                                        const uint32_t imageCountLimit,
+                                                        const uint64_t sizeHint,
```
Nit/optional: Remove "const" from these two parameters.

wantehchang commented on src/read.c:
```diff
@@ -451,6 +475,119 @@ static avifBool avifCodecDecodeInputGetSamples(avifCodecDecodeInput * decodeInpu
+        return AVIF_FALSE;
+    }
+
+    uint8_t layerCount = 0;
```
Nit/optional: This local variable can be declared as `int`.

wantehchang commented on src/read.c:
```diff
@@ -451,6 +475,119 @@ static avifBool avifCodecDecodeInputGetSamples(avifCodecDecodeInput * decodeInpu
+            assert(layerCount == 3);
+            ++layerCount;
+            layerSizes[3] = remainingSize;
+        }
```
Nit/optional: Lines 496 and 518 can be simplified if we declared the layerSize array in avifAV1LayeredImageIndexingProperty to have 4 elements and set the last element layerSize[3] to 0. Then the for loop at line 497 can have 4 iterations (` i < 4` ) and we can delete the if statement at line 514.

wantehchang commented on src/read.c:
```diff
@@ -451,6 +475,119 @@ static avifBool avifCodecDecodeInputGetSamples(avifCodecDecodeInput * decodeInpu
+                                      "lsel property requests layer index [%u] which isn't present in a1lx property ([%u] layers)",
+                                      lselProp->u.lsel.layerID,
+                                      layerCount);
+                return AVIF_FALSE;
```
IMPORTANT: This condition is not an error. It just means we cannot use the optimization. I think lines 531-548 should read:

```
        if (layerCount > 0 && lselProp->u.lsel.layerID < layerCount) {
            // Optimization: If we're selecting a layer that doesn't require the entire image's payload (hinted via the a1lx box)
            for (uint8_t i = 0; i <= lselProp->u.lsel.layerID; ++i) {
                sampleSize += layerSizes[i];
            }
        } else {
            // This layer's payload subsection is unknown, just use the whole payload
            sampleSize = item->size;
        }
```

wantehchang commented on src/read.c:
```diff
@@ -451,6 +475,119 @@ static avifBool avifCodecDecodeInputGetSamples(avifCodecDecodeInput * decodeInpu
+            }
+
+            for (uint8_t i = 0; i <= lselProp->u.lsel.layerID; ++i) {
+                sampleSize += layerSizes[i];
```
[Just a comment, no changes requested] Note that `sampleSize` may be larger than the actual value if the `spatial_id` values are not consecutive, but at least `sampleSize` may be smaller than `item->size`.

wantehchang commented on src/read.c:
```diff
@@ -451,6 +475,119 @@ static avifBool avifCodecDecodeInputGetSamples(avifCodecDecodeInput * decodeInpu
+        assert(lselProp->u.lsel.layerID < MAX_AV1_LAYER_COUNT);
+        sample->offset = 0;
+        sample->size = sampleSize;
+        sample->skip = (uint8_t)lselProp->u.lsel.layerID;
```
IMPORTANT: This is incorrect if the `spatial_id` values are not consecutive.

wantehchang replied:
Since sample->skip is the spatial_id value, we can have the AV1 decoder output all layers and look for a matching spatial_id in the output frames.

- dav1d: check the `frame_hdr->spatial_id` field of the `Dav1dPicture` struct.
- libaom: check the `spatial_id` field of the `aom_image_t` struct. Unfortunately libaom has a bug in this area and I forgot to check in my bug fix: https://aomedia-review.googlesource.com/c/aom/+/133003. Sorry about that.
- libgav1: check the `spatial_id` field of the `Libgav1DecoderBuffer` struct.

wantehchang commented on src/read.c:
```diff
@@ -451,6 +475,119 @@ static avifBool avifCodecDecodeInputGetSamples(avifCodecDecodeInput * decodeInpu
+            sample->offset = offset;
+            sample->size = layerSizes[i];
+            sample->skip = 0;
+            sample->sync = (i == 0); // Assume all layers depend on the first layer
```
Just wanted to confirm that things will still work if we set `sample->sync` to false incorrectly.

wantehchang commented on src/read.c:
```diff
@@ -833,9 +990,17 @@ static avifResult avifDecoderItemRead(avifDecoderItem * item, avifIO * io, avifR
-        totalBytesToRead = partialByteCount;
+    const size_t readOutputSize = (partialByteCount && (partialByteCount < item->size)) ? partialByteCount : item->size;
+
+    // Read in everything from the item's offset 0 all the way to the end of the read
```
Delete the "0" after "the item's offset"?

wantehchang commented on src/read.c:
```diff
@@ -833,9 +990,17 @@ static avifResult avifDecoderItemRead(avifDecoderItem * item, avifIO * io, avifR
-    size_t totalBytesToRead = item->size;
-    if (partialByteCount && (totalBytesToRead > partialByteCount)) {
-        totalBytesToRead = partialByteCount;
+    const size_t readOutputSize = (partialByteCount && (partialByteCount < item->size)) ? partialByteCount : item->size;
```
Should we replace `item->size` with `item->size - offset` (after making sure that `offset < item->size`)?

wantehchang commented on src/read.c:
```diff
@@ -981,11 +1149,21 @@ static avifBool avifDecoderDataGenerateImageGridTiles(avifDecoderData * data, av
+                            decoder->imageCount = tile->input->samples.count;
+                        }
+                    }
+                }
```
Merge the nested ifs:

```
                if (!alpha && item->progressive) {
                    decoder->progressiveState = AVIF_PROGRESSIVE_STATE_AVAILABLE;
                    if (tile->input->samples.count > 1) {
                        decoder->progressiveState = AVIF_PROGRESSIVE_STATE_ACTIVE;
                        decoder->imageCount = tile->input->samples.count;
                    }
                }
```

wantehchang commented on src/read.c:
```diff
@@ -1704,6 +1942,32 @@ static avifBool avifParseItemPropertyAssociation(avifMeta * meta, const uint8_t
+                            return AVIF_FALSE;
+                        }
+                    }
+                }
```
Would you like to check that a1lx is not marked as essential?

>   AVIF: Section 2.3.2.3.2: "If associated, it shall not be marked as essential."

wantehchang commented on src/read.c:
```diff
@@ -2729,6 +2997,37 @@ avifResult avifDecoderParse(avifDecoder * decoder)
+        } else {
+            avifDiagnosticsPrintf(data->diag, "Item ID [%u] is missing a mandatory ispe property", item->id);
+            return AVIF_RESULT_BMFF_PARSE_FAILED;
+        }
```
[Just a comment] I like this change. So far I have had to check if the container width and height is 0 after avifDecoderParse() returns successfully. I will be able to remove those checks in the future.

wantehchang commented on src/scale.c:
```diff
@@ -0,0 +1,122 @@
+#endif
+
+// This should be configurable and/or smarter
+#define AVIF_LIBYUV_FILTER_MODE kFilterBox
```
Add a comment to note that `kFilterBox` has the highest quality (but is also the slowest).

But actually, since the low-resolution frame will be displayed only temporarily, it seems better to use the fastest filter mode.

tongyuantongyu replied:
I believe smooth version is preferred here. libjpeg-turbo even applies extra blur to smooth the decoded image: https://github.com/libjpeg-turbo/libjpeg-turbo/issues/459.

wantehchang commented on src/scale.c:
```diff
@@ -0,0 +1,122 @@
+        return AVIF_FALSE;
+    }
+
+    uint8_t * srcYUVPlanes[AVIF_PLANE_COUNT_YUV];
```
Nit: const pointer:

`const uint8_t * srcYUVPlanes[AVIF_PLANE_COUNT_YUV];`

wantehchang commented on src/scale.c:
```diff
@@ -0,0 +1,122 @@
+    const avifBool srcImageOwnsYUVPlanes = image->imageOwnsYUVPlanes;
+    image->imageOwnsYUVPlanes = AVIF_FALSE;
+
+    uint8_t * srcAlphaPlane = image->alphaPlane;
```
Nit: const pointer:

const uint8_t * srcAlphaPlane = image->alphaPlane;

wantehchang commented on src/scale.c:
```diff
@@ -0,0 +1,122 @@
+        const uint32_t dstUVHeight = (dstHeight + formatInfo.chromaShiftY) >> formatInfo.chromaShiftY;
+
+        for (int i = 0; i < AVIF_PLANE_COUNT_YUV; ++i) {
+            if (!srcYUVPlanes[i]) {
```
Nit/optional: Instead of always doing three iterations and checking if srcYUVPlanes[i] is a null pointer, it would be better to check if `image->yuvFormat == AVIF_PIXEL_FORMAT_YUV400` to determine the number of iterations (1 or 3) of this for loop. See the code in `avifImageAllocatePlanes()`.

wantehchang commented on src/scale.c:
```diff
@@ -0,0 +1,122 @@
+            const uint32_t dstW = (i == AVIF_CHAN_Y) ? dstWidth : dstUVWidth;
+            const uint32_t dstH = (i == AVIF_CHAN_Y) ? dstHeight : dstUVHeight;
+            if (image->depth > 8) {
+                uint16_t * srcPlane = (uint16_t *)srcYUVPlanes[i];
```
Nit: const pointer:

const uint16_t * srcPlane = (const uint16_t *)srcYUVPlanes[i];

wantehchang commented on src/scale.c:
```diff
@@ -0,0 +1,122 @@
+        avifImageAllocatePlanes(image, AVIF_PLANES_A);
+
+        if (image->depth > 8) {
+            uint16_t * srcPlane = (uint16_t *)srcAlphaPlane;
```
Nit: const pointer:

const uint16_t * srcPlane = (const uint16_t *)srcAlphaPlane;

wantehchang commented on src/scale.c:
```diff
@@ -0,0 +1,122 @@
+                uint16_t * dstPlane = (uint16_t *)image->yuvPlanes[i];
+                ScalePlane_12(srcPlane, srcYUVRowBytes[i], srcW, srcH, dstPlane, image->yuvRowBytes[i], dstW, dstH, AVIF_LIBYUV_FILTER_MODE);
+            } else {
+                uint8_t * srcPlane = srcYUVPlanes[i];
```
Nit: const pointer:

const uint8_t * srcPlane = srcYUVPlanes[i];

wantehchang commented on src/scale.c:
```diff
@@ -0,0 +1,122 @@
+            uint16_t * dstPlane = (uint16_t *)image->alphaPlane;
+            ScalePlane_12(srcPlane, srcAlphaRowBytes, srcWidth, srcHeight, dstPlane, image->alphaRowBytes, dstWidth, dstHeight, AVIF_LIBYUV_FILTER_MODE);
+        } else {
+            uint8_t * srcPlane = srcAlphaPlane;
```
Nit: const pointer:

`const uint8_t * srcPlane = srcAlphaPlane;`

wantehchang reviewed (changes requested):
I have reviewed every file. I don't understand the new offset parameter to `avifDecoderItemRead()`. I will try to study that code.

wantehchang commented on src/read.c:
```diff
@@ -451,6 +475,119 @@ static avifBool avifCodecDecodeInputGetSamples(avifCodecDecodeInput * decodeInpu
+        sample->size = sampleSize;
+        sample->skip = (uint8_t)lselProp->u.lsel.layerID;
+        sample->sync = AVIF_TRUE;
+    } else if (allowProgressive && a1lxProp) {
```
Nit: Replace a`1lxProp` with `item->progressive`:

`} else if (allowProgressive && item->progressive) {`

joedrago commented:
@wantehchang I have updated the `lsel` implementation to check `spatial_id` instead of an index-based skip. You said that there was some issue with libaom where we'll need to do a version check, but I'll defer writing the proper #ifdef to you.  I've also added in your changes above.

We also still need to address the scaling code and potential libyuv scaling recontribution.

wantehchang reviewed (approved):
Joe: Thank you for the new commits. I think this PR is ready to be merged. Please add a warning comment that the allowProgressive decoder setting should not be enabled except for experiments, because the implementation still has some issues and there is a small chance that the spec may change based on your implementation experience.
